### ISSUE TITLE
Help: Information not refreshed

### ISSUE THREAD

**Issue by mbolo01** [giachello/beoplay#9]
Hi there,

I need help on the following:

I recently noticed that my B&O HA media_player entities were no longer refreshing the media image of the currently playing track when using Airplay against a B&O product.

I first tried to isolate any changes in several components that could have triggered this issue, O/S change, music player application change, B&O device software change, but I could not find any obvious answer. Issue occurs with any Airplay source (OSX, IOS), including very old ones, it occurs also with different music players and against B&O devices that I kept out from software updates.

The only changes I could think of are HA updates.

I had a deeper look at a given B&O HA entity attributes while playing different tracks from different albums, so with different media image, and I could find that "Entity picture" was not changing while others such as "Volume Level", "Media title", "Media artist", "Media track" were changing properly and accordingly.

The "Entity picture" sticks with the very first value it collected when entity was first activated e.g. first play after entity has been turned off or after a HA restart.

The Entity picture value is in the form of "/api/media_player_proxy/media_player.bs_core?token=fa001c96da44c5d55f68214ddf4cd51111a3316e8ae0581b3d9d94b227611c7f&cache=34af6d1562400fa7"  and this token value stays the same until entity off/on

My  media_player.py "media_image_url" property is set as follow (note: I have two static images depending on the source, otherwise it is dynamic).

@property
    def media_image_url(self):
        """Image url of current playing media."""
        if self._source == "Optical":
            return "http://192.168.1.52:8123/local/TVFrame.jpg"
        elif self._source == "Line-In":
            return "http://192.168.1.52:8123/local/BSMoment.jpeg"
        return self._media_url if self._media_url else None 

Thanks in advance for any help you can provide.
Kind regards.
Jérôme

**Comment by mbolo01** [comment]
More info: looking at the product notification output, track image URL has always the same value for all played tracks for the duration of the streaming up to some kind of inactivity timeout (I guess), e.g.:

{"notification":{"timestamp":"2021-09-07T14:13:51.769715","type":"NOW_PLAYING_STORED_MUSIC","kind":"playing","data":{"name":"Temps √† nouveau ","duration":287,"trackImage":[{"url":"**http://xxx.xxx.xxx:80/api/stream//tmp/temp_data_airPlayAlbum_87101b87cb47293c95d54f431b2405e8**","size":"large","mediatype":"image/jpg"}],"artist":"Jean-Louis Aubert","artistImage":[],"album":"Deux pages de tourn√©e","albumImage":[],"genre":"","playQueueItemId":"airplay"}}}
"airplay"}}}

But launching the URL changes the content according to what is playing.

For me it looks like the beoplay entity does not read the URL again as there is no change to it but instead exposes the contente (image) that was retrieved at first read (and cached ?) for the duration of the stream and as long as the URL does not change, hopefully my explanations are clear enough.

**Comment by giachello** [comment]
hi there, this is very helpful. To be honest, I think this plugin needs a thorough scrub. There are a lot of special code that do weird things in it and a clean rewrite would  help a lot to iron out these kinks. The Spotify integration also has these special images so it may be a matter of copying the way they do it.

Unfortunately, I don't have the time right now to devote to this :-( but maybe later in the Fall as things quiet down...

Giovanni

**Comment by mbolo01** [comment]
Thanks for the feedback, and no worries, I'll wait!

**Comment by riverstyx99** [comment]
Hi Jérôme,

Perhaps as a short term solution you may be able work around this issue by appending:
"?foo=HHMMSS" or similar to the end of the image URL. (or "&foo=HHMMSS" if the URL already incorporates some GET method parameters as I know some products / sources do).

I just tried grabbing a station image for netradio on my Avant with an extra parameter added and it served the image without complaint, and hopefully the fact that the URL has changed would be enough to bypass any caching.

Martin.

**Comment by mbolo01** [comment]
Hi Martin,

Thanks for the proposed trick.

I have no issues with radios, nor with Chromecast, any station/track changes triggers corresponding image change in the media_player card. It only happened recently with Airplay, without any change that I could remember on the transmitter side (e.g. iPad IOS has not changed, happens with any music application when Airplay is used) or the receiver side (I always keep a B&O speaker with  software at version N-1 and all my speakers have the same issue). The only changes I'm sure about are Home Assistant code and dependencies.

The URL I'm referring to in my previous post is internal to the B&O device, nothing I can change here (at least with my basic tech knowledge). 

My analysis is that when the B&O device starts a new Airplay session, it sets a pointer to a temp area where the image received via Airplay is stored and reread its content upon track change. When you load this unique URL pointing to the B&O device temp area through a desktop browser, each time track changes, image changes accordingly. Home Assistant entity associates a token to this URL and I guess caches the content until the URL changes, where it should reread the content whenever a track change occur.

Jérôme

**Comment by riverstyx99** [comment]
Hi Jérôme,

Please try the following changes in media_player.py - as I mentioned, I don't use airplay so I haven't been able to test whether this fixes your issue, but it doesn't appear to break fetching cover art for any of the other sources I use on my B&O devices.

Add the following up near the top with the other import statements:
`
import urllib.request
`

Then change the media_image_url section to:
`
    @property
    def media_image_url(self):
        """Image url of current playing media."""
        if self._source == "Optical":
            return "http://192.168.1.52:8123/local/TVFrame.jpg"
        elif self._source == "Line-In":
            return "http://192.168.1.52:8123/local/BSMoment.jpeg"
        if self._media_url and self._media_track:
            if "?" in self._media_url:
                return self._media_url + "&track=" + urllib.request.pathname2url(self._media_track)
            else:
                return self._media_url + "?track=" + urllib.request.pathname2url(self._media_track)
    return self._media_url if self._media_url else None
`

What this does, to use the example data in your earlier post, is to modify the URL from:

`http://xxx.xxx.xxx:80/api/stream//tmp/temp_data_airPlayAlbum_87101b87cb47293c95d54f431b2405e8`

to:

`http://xxx.xxx.xxx:80/api/stream//tmp/temp_data_airPlayAlbum_87101b87cb47293c95d54f431b2405e8?track=Temps%20%E2%88%9A%E2%80%A0%20nouveau%20`

The extra part of the url is not expected by the B&O device and is therefore ignored (it still serves the correct image), but the fact that the URL has changed should be enough to resolve the caching issue.

My original suggestion of appending a timestamp should also work, but this solution is perhaps a little more elegant as it only resets the cached image when the currently playing track changes, rather than at set time intervals.

The issue only seems to occur on Airplay, because the URL doesn't change when the image changes, whereas on other sources the URL changes for each image.

As giachello mentioned, there may be a way to flush the cache, but it may not be straightforward to determine when this should happen, so doing it this way at least means the cached image is only discarded when the track title has changed

Please let me know how you get on...
Martin.

**Comment by mbolo01** [comment]
Hi Martin,

You are a star! Your solution works very well with Airplay, image is changed properly now, and there is not impact on other sources (even if I had no doubt, its always better to test).  
Many thanks for that, very clever solution until a cache flush is introduced.

Cheers,
Jérôme

**Comment by riverstyx99** [comment]
Thanks Jérôme,

Glad to hear it worked for you.

There's a very obvious edge case where this workaround won't work, and that is where there are two tracks with the same name played consecutively (maybe by different artists, or by the same artist but from a different album). Given it is usually one image for per album, it might be better implemented using the same strategy, but using artist and album names instead (or some computed hash of those).

I'm currently working on pybeoplay though, to try and grab metadata from some of the sources that currently don't show it, and to resolve the issue of stale metadata when a source is changed, so I'll come back to this issue at a later date.

Martin.

**Comment by giachello** [comment]
@riverstyx99  Please integrate your code back into pybeoplay when you are done! Would love to include this in the default distribution. Will update the pypi package when you are done.

**Comment by riverstyx99** [comment]
Thanks @giachello

I've submitted a pull request for the changes I've made to pybeoplay thus far. I'm not reliant on pypi being updated as I have installed a local copy of the package in /config/pybeoplay and HA seems happy to use this in preference to the pypi one, but I know others will be using it. It's up to you how often you want to push new versions to pypi, but there's no pressure from my end, and I may send you some more changes over the coming days. I'll tend towards submitting a pull requests each time I've finished work on a given issue, as I think that makes the code changes easier to review.

I don't have anything hardcoded in my code, they were just the two image URLs that mbolo01 had included in his original post, so I included them again in my example :)

Cheers,
Martin.
